nilearn/input_data/__init__.py

```diff
@@ -1,12 +1,31 @@
 """
-The :mod:`nilearn.input_data` module includes scikit-learn tranformers.
+The :mod:`nilearn.input_data` module used to include masker objects.
+It is deprecated since release 0.9.0 in favor of the
+:mod:`~nilearn.maskers` module.
+
+Please consider updating your code:
+
+.. code-blocks::python
+    from nilearn.input_data import NiftiMasker
+
+becomes:
+
+.. code-blocks::python
+    from nilearn.maskers import NiftiMasker
+
+Note that all imports that used to work will continue to do so with
+a simple warning at least until release 0.13.0.
 """
-from .nifti_masker import NiftiMasker
-from .multi_nifti_masker import MultiNiftiMasker
-from .nifti_labels_masker import NiftiLabelsMasker
-from .nifti_maps_masker import NiftiMapsMasker
-from .nifti_spheres_masker import NiftiSpheresMasker
+import sys
+import warnings
+
+from nilearn import maskers  # noqa:F401
+
+warnings.warn("The module 'input_data' is deprecated since 0.9.0, "
+              "Importing maskers from 'input_data' will be possible "
+              "at least until release 0.13.0. Please import maskers "
+              "from the 'maskers' module instead.",
+              FutureWarning)
 
+sys.modules[__name__] = sys.modules['nilearn.maskers']
```

Comment:
this makes sense and from the documentation on the import system and in particular [this part](https://docs.python.org/3/reference/import.html#loading) seems like it should work, and in practice seems to give the desired result at least on my machine! I'm just wondering if there are traps we're not aware of; by any chance do you have an example of this solution being used in another project, or suggested in some resource?

(note of course the backwards compatibility will not be 100% perfect, eg the modules' `__name__` and objects' `__module__` will change, but I think that's acceptable)

Reply:
I don't know if this solution is used in other projects. I'll take a look around! :eyes:

Reply:
@jeromedockes I've asked around on the PARIETAL engineers channel, but got very few answers.
Guillaume pointed me to this PR on scikit-learn: https://github.com/scikit-learn/scikit-learn/pull/14939 where they had to do something similar. It is not exactly the same though, as they don't rename the module (meaning the name of the folder), but the files themselves to prepend them with "_". They basically keep two versions of each file "file.py" and "_file.py" but the former imports from the latter with some deprecation warnings.
Our use case is slightly different since we keep all files, but move them to another folder. We could have the same mechanism, but I don't think it is necessary.
WDYT?

Reply:
Yes I agree the situation is a  bit different. we can try it as it is in the PR then. we get some slightly awkward things like the output of the script below but I guess that's unavoidable
```python
from nilearn.input_data import NiftiMapsMasker as masker1
from nilearn.input_data.nifti_maps_masker import NiftiMapsMasker as masker2

print(masker1 is masker2)
print(masker1.__module__)
print(masker2.__module__)
```

```
False
nilearn.maskers.nifti_maps_masker
nilearn.input_data.nifti_maps_masker
```

Reply:
@jeromedockes I've implemented a possible alternative solution inspired by the PR from scikit-learn above (see commit d2fbbe8). This should solve the issue in your last comment. However, I'm still unsure whether or not this is a better solution. 

Potential issues:

- You cannot import private objects from `input_data`. So all user code doing things like `from nilearn.input_data.nifti_masker import _get_mask_strategy` will break. You can argue that this is a risk you're willing to take if you import private things, but we could also import them manually in each python file in `input_data` to be backward compatible even for private objects.
- When you import `input_data` the first time, you get 5 warnings due to the imports in `__init__.py`. This is not a big issue and there might be ways to avoid this. I can take a closer look if we prefer this solution over the first one. 

Also, I realized a few things:

- `nilearn.input_data.masker_validation.check_embedded_nifti_masker` looks like it is public, but it is not documented in the API. Should we make it private or public?
- Same comment for the content of `base_masker.py` (`BaseMasker` class and `filter_and_extract` function)

Reply:
Hi, thanks for looking at this! Re. the potential issues:

- even though nilearn documentation doesn't explicitly state we follow it, the convention that `_` names are for internal/private objects is used everywhere. client code shouldn't use them and we don't need to avoid breaking code that does.
- for the warnings when importing `input_data`, could we just put one warning in `input_data.__init__` and remove those in the submodules?

That being said if you prefer the first solution modifying `sys.modules`, I'm ok with it!

> Also, I realized a few things:

IMO these should be private, they are clearly internals of nilearn's estimators and maskers and it seems unlikely users would use them, especially as they are undocumented and not listed in `__all__`. Except perhaps the `BaseMasker` if we think it's possible for users to inherit from it to create their own maskers